File: plugins/moe/.opencode/skills/extracting-requirements/scripts/validate_scenarios.py
#!/usr/bin/env python3
"""Validate behavior-scenarios.md format and cross-references.

Usage: validate_scenarios.py <scenarios-file> <requirements-dir>

Checks:
- Every scenario has required fields (kind, proof seam, owning stories)
- All owning story references resolve to existing STORY-IDs
- Journey scenarios have steps (not empty)
- Scenario IDs are unique
- No UNRESOLVED() story references remain
"""
import re
import sys
from pathlib import Path


def load_story_ids(requirements_dir: Path) -> set[str]:
    """Collect all STORY-NNNN IDs from per-epic files."""
    ids: set[str] = set()
    for epic_file in requirements_dir.glob("EPIC-*.md"):
        for line in epic_file.read_text().splitlines():
            m = re.match(r"^## (STORY-\d+)", line)
            if m:
                ids.add(m.group(1))
    return ids


def validate(scenarios_path: Path, requirements_dir: Path) -> list[str]:
    """Validate scenarios file. Returns list of error strings."""
    errors: list[str] = []
    text = scenarios_path.read_text()
    lines = text.splitlines()

    known_stories = load_story_ids(requirements_dir)
    seen_ids: set[str] = set()

    current_id = None
    current_kind = None
    has_steps = False
    has_owning = False
    has_seam = False

    def flush():
        nonlocal current_id, current_kind, has_steps, has_owning, has_seam
        if current_id is None:
            return
        if not has_owning:
            errors.append(f"{current_id}: missing 'Owning stories' field")
        if not has_seam:
            errors.append(f"{current_id}: missing 'Proof seam' field")
        if current_kind == "journey" and not has_steps:
            errors.append(f"{current_id}: journey scenario has no steps")
        current_id = None
        current_kind = None
        has_steps = False
        has_owning = False
        has_seam = False

    for i, line in enumerate(lines, 1):
        # Scenario header
        id_match = re.match(r"^## (SCENARIO-\d+|JOURNEY-\d+)", line)
        if id_match:
            flush()
            current_id = id_match.group(1)
            if current_id in seen_ids:
                errors.append(f"line {i}: duplicate scenario ID {current_id}")
            seen_ids.add(current_id)

        if current_id:
            if line.startswith("**Kind:**"):
                current_kind = line.split(":**")[1].strip().lower()
            if line.startswith("**Proof seam:**"):
                has_seam = True
            if line.startswith("**Owning stories:**"):
                has_owning = True
                refs = line.split(":**")[1].strip()
                for ref in re.findall(r"STORY-\d+", refs):
                    if ref not in known_stories:
                        errors.append(f"{current_id}: references unknown {ref}")
                for unresolved in re.findall(r"UNRESOLVED\([^)]+\)", refs):
                    errors.append(f"{current_id}: has {unresolved}")
            if re.match(r"^\d+\.", line.strip()):
                has_steps = True

    flush()

    if not seen_ids:
        errors.append("no scenarios found in file")

    return errors


def main() -> int:
    if len(sys.argv) != 3:
        print(f"usage: {sys.argv[0]} <scenarios-file> <requirements-dir>",
              file=sys.stderr)
        return 2

    scenarios_path = Path(sys.argv[1])
    requirements_dir = Path(sys.argv[2])

    if not scenarios_path.exists():
        print(f"error: file not found: {scenarios_path}", file=sys.stderr)
        return 2
    if not requirements_dir.is_dir():
        print(f"error: directory not found: {requirements_dir}", file=sys.stderr)
        return 2

    errors = validate(scenarios_path, requirements_dir)
    if errors:
        for e in errors:
            print(f"ERROR: {e}", file=sys.stderr)
        print(f"FAIL: {len(errors)} error(s)")
        return 1

    print("OK: scenarios valid")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: packages/core/skills/extracting-requirements/scripts/validate_scenarios.mjs
import { existsSync, readFileSync, readdirSync, statSync } from "node:fs";
import { join, resolve, sep } from "node:path";
import { pathToFileURL } from "node:url";

function pathParts(path, separatorPattern) {
  return path.split(separatorPattern).filter((part) => part && part !== ".");
}

function normalizeWindowsPathSpelling(path) {
  const windowsPath = path.replaceAll("/", "\\");
  const unc = windowsPath.match(/^\\\\([^\\]+)\\([^\\]+)(?:\\|$)/);
  if (unc) {
    const root = `\\\\${unc[1]}\\${unc[2]}\\`;
    const parts = pathParts(windowsPath.slice(unc[0].length), /\\+/);
    return parts.length > 0 ? `${root}${parts.join("\\")}` : root;
  }
  const drive = windowsPath.match(/^([A-Za-z]:)(\\?)/);
  if (drive) {
    const rooted = drive[2] === "\\";
    const parts = pathParts(windowsPath.slice(drive[1].length + (rooted ? 1 : 0)), /\\+/);
    if (rooted) return parts.length > 0 ? `${drive[1]}\\${parts.join("\\")}` : `${drive[1]}\\`;
    return parts.length > 0 ? `${drive[1]}${parts.join("\\")}` : drive[1];
  }
  const rooted = windowsPath.startsWith("\\");
  const parts = pathParts(windowsPath.slice(rooted ? 1 : 0), /\\+/);
  if (rooted) return parts.length > 0 ? `\\${parts.join("\\")}` : "\\";
  return parts.length > 0 ? parts.join("\\") : ".";
}

function normalizePathSpelling(path) {
  if (sep === "\\") return normalizeWindowsPathSpelling(path);
  const doubleSlashRoot = path.startsWith("//") && !path.startsWith("///");
  const root = doubleSlashRoot ? "//" : path.startsWith("/") ? "/" : "";
  const parts = pathParts(path, /\/+/);
  if (root) return parts.length > 0 ? `${root}${parts.join("/")}` : root;
  return parts.length > 0 ? parts.join("/") : ".";
}

function compareCodePoints(left, right) {
  const leftPoints = Array.from(left, (character) => character.codePointAt(0));
  const rightPoints = Array.from(right, (character) => character.codePointAt(0));
  const length = Math.min(leftPoints.length, rightPoints.length);
  for (let index = 0; index < length; index += 1) {
    if (leftPoints[index] !== rightPoints[index]) return leftPoints[index] - rightPoints[index];
  }
  return leftPoints.length - rightPoints.length;
}

function splitLines(text) {
  return text.split(/\r\n|[\n\r\v\f\x1c-\x1e\x85\u2028\u2029]/);
}

export function loadStoryIds(requirementsDirectory) {
  const ids = new Set();
  const epicFiles = readdirSync(requirementsDirectory)
    .filter((name) => /^EPIC-.*\.md$/.test(name))
    .sort(compareCodePoints);

  for (const name of epicFiles) {
    for (const line of splitLines(readFileSync(join(requirementsDirectory, name), "utf8"))) {
      const match = line.match(/^## (STORY-\p{Nd}+)/u);
      if (match) ids.add(match[1]);
    }
  }
  return ids;
}

export function validateScenarios(scenariosPath, requirementsDirectory) {
  const errors = [];
  const lines = splitLines(readFileSync(scenariosPath, "utf8"));
  const knownStories = loadStoryIds(requirementsDirectory);
  const seenIds = new Set();

  let currentId;
  let currentKind;
  let hasSteps = false;
  let hasOwning = false;
  let hasSeam = false;

  function flush() {
    if (currentId === undefined) return;
    if (!hasOwning) errors.push(`${currentId}: missing 'Owning stories' field`);
    if (!hasSeam) errors.push(`${currentId}: missing 'Proof seam' field`);
    if (currentKind === "journey" && !hasSteps) {
      errors.push(`${currentId}: journey scenario has no steps`);
    }
    currentId = undefined;
    currentKind = undefined;
    hasSteps = false;
    hasOwning = false;
    hasSeam = false;
  }

  for (const [index, line] of lines.entries()) {
    const idMatch = line.match(/^## (SCENARIO-\p{Nd}+|JOURNEY-\p{Nd}+)/u);
    if (idMatch) {
      flush();
      currentId = idMatch[1];
      if (seenIds.has(currentId)) {
        errors.push(`line ${index + 1}: duplicate scenario ID ${currentId}`);
      }
      seenIds.add(currentId);
    }

    if (currentId !== undefined) {
      if (line.startsWith("**Kind:**")) {
        currentKind = line.split(":**")[1].trim().toLowerCase();
      }
      if (line.startsWith("**Proof seam:**")) hasSeam = true;
      if (line.startsWith("**Owning stories:**")) {
        hasOwning = true;
        const references = line.split(":**")[1].trim();
        for (const match of references.matchAll(/STORY-\p{Nd}+/gu)) {
          if (!knownStories.has(match[0])) {
            errors.push(`${currentId}: references unknown ${match[0]}`);
          }
        }
        for (const match of references.matchAll(/UNRESOLVED\([^)]+\)/g)) {
          errors.push(`${currentId}: has ${match[0]}`);
        }
      }
      if (/^\p{Nd}+\./u.test(line.trim())) hasSteps = true;
    }
  }

  flush();
  if (seenIds.size === 0) errors.push("no scenarios found in file");
  return errors;
}

export const validate = validateScenarios;

export function main(args = process.argv.slice(2)) {
  if (args.length !== 2) {
    process.stderr.write(`usage: ${process.argv[1]} <scenarios-file> <requirements-dir>\n`);
    return 2;
  }

  const scenariosPath = normalizePathSpelling(args[0]);
  const requirementsDirectory = normalizePathSpelling(args[1]);
  if (!existsSync(scenariosPath)) {
    process.stderr.write(`error: file not found: ${scenariosPath}\n`);
    return 2;
  }
  if (!existsSync(requirementsDirectory) || !statSync(requirementsDirectory).isDirectory()) {
    process.stderr.write(`error: directory not found: ${requirementsDirectory}\n`);
    return 2;
  }

  const errors = validateScenarios(scenariosPath, requirementsDirectory);
  if (errors.length > 0) {
    for (const error of errors) process.stderr.write(`ERROR: ${error}\n`);
    process.stdout.write(`FAIL: ${errors.length} error(s)\n`);
    return 1;
  }

  process.stdout.write("OK: scenarios valid\n");
  return 0;
}

const isDirect = process.argv[1] && pathToFileURL(resolve(process.argv[1])).href === import.meta.url;
if (isDirect) process.exitCode = main();
